doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,555 @@
+can choose to not mutate by simply creating a new buffer with a copy of the data
+in it. A return value of 1 indicates that the callback completed successfully. A
+return value of 0 indicates a fatal error.
+
+Once libssl has finished using the mutated buffer it will call the
+`finish_mutate_handshake_cb` callback which can then release the buffer and
+perform any other cleanup as required.
+
+### QUIC Pre-Encryption Packets
+
+QUIC Packets are the primary mechanism for exchanging protocol data within QUIC.
+Multiple packets may be held within a single datagram, and each packet may
+itself contain multiple frames. A packet gets protected via an AEAD encryption
+algorithm prior to it being sent. Fault Injector based tests may need to inject
+faults into these packets prior to them being encrypted.
+
+An example test might insert an unrecognised frame type into a QUIC packet to
+confirm that an OpenSSL QUIC client handles it appropriately (e.g. by raising a
+protocol error).
+
+The above functionality will be supported by the following two new callbacks
+which will provide the ability to mutate packets before they are encrypted and
+sent. As for the TLS callbacks these will be internal only and not part of the
+public API.
+
+```` C
+typedef int (*ossl_mutate_packet_cb)(const QUIC_PKT_HDR *hdrin,
+                                     const OSSL_QTX_IOVEC *iovecin, size_t numin,
+                                     QUIC_PKT_HDR **hdrout,
+                                     const OSSL_QTX_IOVEC **iovecout,
+                                     size_t *numout,
+                                     void *arg);
+
+typedef void (*ossl_finish_mutate_cb)(void *arg);
+
+void ossl_qtx_set_mutator(OSSL_QTX *qtx, ossl_mutate_packet_cb mutatecb,
+                          ossl_finish_mutate_cb finishmutatecb, void *mutatearg);
+````
+
+A single new function call will set both callbacks. The `mutatecb` callback will
+be invoked after a packet has been constructued but before protection has
+been applied to it. The header for the packet will be pointed to by `hdrin` and
+the payload will be in an iovec array pointed to by `iovecin` and containing
+`numin` iovecs. The `mutatecb` callback is expected to allocate a new header
+structure and return it in `*hdrout` and a new set of iovecs to be stored in
+`*iovecout`. The number of iovecs need not be the same as the input. The number
+of iovecs in the output array is stored in `*numout`. Optionally the callback
+can choose to not mutate by simply creating new iovecs/headers with a copy of the
+data in it. A return value of 1 indicates that the callback completed
+successfully. A return value of 0 indicates a fatal error.
+
+Once the OpenSSL QUIC implementation has finished using the mutated buffers the
+`finishmutatecb` callback is called. This is expected to free any resources and
+buffers that were allocated as part of the `mutatecb` call.
```

Comment:
I assume this mutator API does not allow you to insert extra packets into the traffic or completely drop a packet. Do we need this functionality?

Reply:
Correct it does not currently allow that. I think we could easily extend it in the future if we need that.